core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,307 @@
+    }
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, this.config.targetFolder);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, this.config.nodePath);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(XMLInputFormat.class, conf)));
+  }
+
+  /**
+   * Method to set file tracking information to temporary file to read.
+   * This is to read / write file tracking information from / to XMLInputFormat and XMLRecordReader using temporary
+   * file, as there is no direct way to read / write information from / to dataset.
+   * @param context
+   * @param conf
+   * @throws IOException
+   */
+  private void setFileTrackingInfo(BatchSourceContext context, Configuration conf) throws IOException {
+    String tableName = this.config.tableName;
+    if (StringUtils.isNotEmpty(tableName)) {
+      try {
+        processedFileTrackingTable = context.getDataset(tableName);
+        if (processedFileTrackingTable != null) {
+          File tempFile = new File(tableName);
+          // if file exists, then delete it and create new.
+          if (tempFile.exists()) {
+            tempFile.delete();
+          }
+          tempFile.createNewFile();
+          FileWriter fw = new FileWriter(tempFile);
+          BufferedWriter bw = new BufferedWriter(fw);
+          CloseableIterator<KeyValue<byte[], byte[]>> iterator = processedFileTrackingTable.scan(null, null);
+
+          //write file tracking information to temporary file.
+          while (iterator.hasNext()) {
+            KeyValue<byte[], byte[]> keyValue = iterator.next();
+            bw.write(new String(keyValue.getKey(), Charsets.UTF_8) + "\n");
+          }
+          bw.close();
+          conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE, tableName);
+        }
+      } catch (IOException ioException) {
+        throw ioException;
+      } catch (DatasetInstantiationException exception) {
+        throw exception;
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    Map<String, String> xmlRecord = (Map<String, String>) input.getValue();
+    String fileName = "";
+    String record = "";
+    for (String key : xmlRecord.keySet()) {
+      fileName = key;
+      record = xmlRecord.get(key);
+    }
+
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_XML_SCHEMA)
+      .set("offset", input.getKey().get())
+      .set("fileName", fileName)
+      .set("record", record)
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    super.onRunFinish(succeeded, context);
+    if (StringUtils.isNotEmpty(this.config.tableName)) {
+      try {
+        //read file tracking information updated by XMLRecordReader and put it in dataset.
```

Comment:
We have dataset access in mapper and reduce tasks (as well as in beforeSubmit/onFinish of mapreduce). However, we don't expose any way for users to get it if they define an InputFormat or from an implementation of RecordReader.

I see two potential options:
1. Write the values to the Configuration object, as @nitinmotgi suggested. However, I'm not sure if the hadoop framework supports updates on the Configuration back to the map reduce driver (so that these updates are visible in the onRunFinish), so this would need to be tested, especially in distributed CDAP/MR.
2. Write to HDFS, instead of a local file, so that the onRunFinish can read the file.